src/network/Connection.ts

```diff
@@ -41,7 +41,7 @@ abstract class Writer extends EventEmitter {
 
     abstract write(message: ClientMessage, resolver: DeferredPromise<void>): void;
 
-    abstract close(): void;
+    abstract close(cause: Error | null): void;
```

Comment:
When can this be null?

Reply:
And, if this can't be null, then I guess we won't need the `closed` boolean anymore

Reply:
Right, we check if it is null and create a new error if it is null in `Connection.close`. Let me try to get rid of `closed` boolean as well.

Reply:
Removed `closed` boolean. Add `isClosed()` method just for readability.